### PR TITLE
Upgrade pybind

### LINKED ISSUE #8944: [Speed] Upgrade pybind from 2.1 to 2.2
To use `py::call_guard`, pybind11 version needs to be at least 2.2.

The current implementation of python `exe.run` will hold the python GIL and blocks other python threads. Because by default, "When calling a C++ function from Python, the GIL is always held."

One way to release the lock is through `py::call_guard<py::gil_scoped_release>()`.

```
m.def("call_go", &call_go, py::call_guard<py::gil_scoped_release>());
```

Links: 
1. pybind11 official documentation: https://github.com/pybind/pybind11/blob/master/docs/advanced/misc.rst#global-interpreter-lock-gil
1. Usage in other project: 
    1. Pytorch: https://github.com/rsumner31/pytorch/blob/d7a6b2cd1b47793379a42d5b133255c28ae3499c/torch/csrc/cuda/python_comm.cpp#L8-L16

### PR DESCRIPTION
**Pull request by tonyyang-svail**
fix #8944

### CODE REVIEW

File: paddle/fluid/pybind/protobuf.cc [hunk 1]
```diff
@@ -21,77 +21,6 @@ limitations under the License. */
 #include "paddle/fluid/framework/program_desc.h"
 #include "paddle/fluid/framework/var_desc.h"
 
-// Cast boost::variant for PyBind.
```
Comment (tonyyang-svail) [inline paddle/fluid/pybind/protobuf.cc]: `boost::variant` compatibility has been added at pybind 2.2

File: paddle/fluid/pybind/CMakeLists.txt [hunk 2]
```diff
@@ -4,6 +4,6 @@ if(WITH_PYTHON)
     DEPS pybind python backward proto_desc paddle_memory executor prune init profiler feed_fetch_method
     ${GLOB_OP_LIB})
   if(NOT APPLE AND NOT ANDROID)
-    target_link_libraries(paddle_pybind rt)
+    target_link_libraries(paddle_pybind rt pybind)
```
Comment (tonyyang-svail) [inline paddle/fluid/pybind/CMakeLists.txt]: `pybind` should be linked to `paddle_pybind`

File: paddle/fluid/pybind/pybind.cc [hunk 3]
```diff
@@ -57,8 +57,8 @@ bool IsCompiledWithCUDA() {
 #endif
 }
 
-PYBIND11_PLUGIN(core) {
-  py::module m("core", "C++ core of PaddlePaddle");
+PYBIND11_MODULE(core, m) {
+  m.doc() = ("C++ core of PaddlePaddle");
```
Comment (tonyyang-svail) [inline paddle/fluid/pybind/pybind.cc]: API changes

File: paddle/fluid/pybind/protobuf.cc (hunk 1 above)
Comment (qingqing01) [reply paddle/fluid/pybind/protobuf.cc]: From the doc, http://pybind11.readthedocs.io/en/latest/advanced/cast/stl.html?highlight=variant#c-17-library-containers  , this feature requires Boost 1.56 or newer. But I ever tried in https://github.com/PaddlePaddle/Paddle/pull/7563, it seems there is some problem. You can try again.

**Comment by typhoonzero** [comment]
Maybe related to 

https://github.com/PaddlePaddle/Paddle/issues/7851

https://github.com/PaddlePaddle/Paddle/issues/7369

**Pull request closed without merging**